openmicroscopy/devspace
# round3

This PR:
 - creates new nginx container for jenkins
 - make all basic services extendable


To test:
 - create VM image from volume (openstack ip 10.0.51.???)
 - choose `m2.large` as you will need RAM

**either use anislbe playbook https://github.com/aleksandra-tarkowska/infrastructure/blob/devspace/openstack/examples/devspace.yml (you will still have to manually add snoopy ssh keys and gitconfig) or manually follow the steps below:**

  - install docker https://docs.docker.com/engine/installation/linux/centos/#/install-with-the-script

 ```
sudo yum update
curl -fsSL https://get.docker.com/ | sh
sudo systemctl enable docker
sudo systemctl start docker
```
 - add centos user to docker group see https://docs.docker.com/engine/installation/linux/centos/#/create-a-docker-group `sudo usermod -aG docker centos`
 - manually modifying default MTU for docker-containers

 ```
$ cp /usr/lib/systemd/system/docker.service to /etc/systemd/system
$ ip addr show eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 ...
# specify “–mtu=1400”
$ vi /etc/systemd/system docker.service
ExecStart=/usr/bin/dockerd --mtu 1400
$ sudo systemctl daemon-reload
$ sudo systemctl restart docker
```

 - it is recomanded to create new user

 ```
sudo adduser omero
sudo usermod -aG docker omero
```

 - install `sudo yum install git python-pip pyth-setuptools python-virtualenv`
 - install compose

 ```
virtualenv --system-site-packages ~/venv
source ~/venv/bin/activate
pip install --upgrade pip setuptools
pip install "requests<2.8,>=2.6.1" docker-py docker-compose
```

 - deploy devspace
 ```
git clone -b nginx_redirect https://github.com/aleksandra-tarkowska/devspace.git
cd devspace/
./rename.py develop
sed -i 's/ARG USER_ID=1000/ARG USER_ID='"$UID"'/g' slave/Dockerfile
sed -i 's/ARG USER_ID=1000/ARG USER_ID='"$UID"'/g' nginx/Dockerfile
sed -i 's/ARG USER_ID=1000/ARG USER_ID='"$UID"'/g' web/Dockerfile
sed -i 's/ARG USER_ID=1000/ARG USER_ID='"$UID"'/g' server/Dockerfile
./sslcert jenkins/sslcert 10.0.51.???
./sslcert nginx/sslcert 10.0.51.???
cp -r ~/snoopy_keys/.ssh ~/devspace/slave/
cp -r ~/snoopy_keys/.gitconfig ~/devspace/slave/
ssh-keyscan github.com >> ~/devspace/slave/.ssh/known_hosts
# check if keys in .ssh has correct permissions, otherwise push jobs will fail
chmod 700 -R ~/devspace/slave/.ssh/
./ds build
./ds up -d
 ```

 - jenkins is run on `https://host:8443`
 - build using Trigger
 - web is run using omero default config on `http://host/web/` **( you have to run first jobs to get web available)**


cc: @manics @joshmoore
